src/routes/webhooks/paddle.ts

```diff
@@ -460,18 +476,325 @@ export const processGiftedPayment = async ({
   );
 };
 
+const checkTransactionStatusValid = ({
+  event,
+  transaction,
+  nextStatus,
+  validStatus,
+  data,
+}: {
+  event: EventEntity;
+  transaction: UserTransaction;
+  nextStatus: UserTransactionStatus;
+  validStatus: UserTransactionStatus[];
+  data: ReturnType<typeof getPaddleTransactionData>;
+}): boolean => {
+  if (!validStatus.includes(transaction.status)) {
+    logger.warn(
+      {
+        eventType: event.eventType,
+        provider: SubscriptionProvider.Paddle,
+        currentStatus: transaction.status,
+        nextStatus,
+        data,
+      },
+      'Transaction with invalid status',
+    );
+
+    return false;
+  }
+
+  return true;
+};
+
 export const processTransactionCompleted = async ({
   event,
 }: {
   event: TransactionCompletedEvent;
 }) => {
+  await notifyNewPaddleTransaction({ event });
+
+  if (isCoreTransaction({ event })) {
+    const transactionData = getPaddleTransactionData({ event });
+    const con = await createOrGetConnection();
+
+    const transaction = await getTransactionForProviderId({
+      con,
+      providerId: transactionData.id,
+    });
+
+    await con.transaction(async (entityManager) => {
+      const userTransaction = await updateUserTransaction({
+        con: entityManager,
+        transaction,
+        nextStatus: UserTransactionStatus.Success,
+        data: transactionData,
+        event,
+      });
+
+      await purchaseCores({
+        transaction: userTransaction,
+      });
+    });
+
+    return;
+  }
+
   const { gifter_id } = (event?.data?.customData ?? {}) as PaddleCustomData;
 
   if (gifter_id) {
     await processGiftedPayment({ event });
   }
+};
 
-  await notifyNewPaddleTransaction({ event });
+export const updateUserTransaction = async ({
+  con,
+  transaction,
+  nextStatus,
+  data,
+}: {
+  con: DataSource | EntityManager;
+  transaction: UserTransaction | null;
+  nextStatus?: UserTransactionStatus;
+  data: ReturnType<typeof getPaddleTransactionData>;
+  event: EventEntity;
+}): Promise<UserTransaction> => {
+  const providerTransactionId = data.id;
+
+  const itemData = data.items[0];
+
+  if (transaction) {
+    if (transaction.receiverId !== data.customData.user_id) {
+      throw new Error('Transaction receiver does not match user ID');
+    }
+
+    if (
+      transaction.status === UserTransactionStatus.Success &&
+      transaction.value !== itemData.price.customData.cores
+    ) {
+      throw new Error('Transaction value changed after success');
+    }
+  }
+
+  const payload = con.getRepository(UserTransaction).create({
+    processor: UserTransactionProcessor.Paddle,
+    id: transaction?.id,
+    receiverId: data.customData.user_id,
+    status: nextStatus,
+    productId: null, // no product user is buying cores directly
```

Comment:
We don't want this as fixed something for future?

Reply:
I though of adding generic `Product` for cores, but since Product entities are something purchased with cores it made more sense to leave `null`, wdyt?

Reply:
Yeah let's leave null then for now